File: LeetCode/636.Exclusive_Time_of_Functions/Mario/636.Exclusive_Time_of_Functions.py
#!/usr/bin/env python3


class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        res = [0] * n
        stack = []  # function_id
        prev_time = 0

        for i, log in enumerate(logs):
            fid, typ, time = log.split(":")
            fid, time = int(fid), int(time)

            if typ == "start":
                if stack:
                    res[stack[-1]] += time - prev_time
                stack.append(fid)
                prev_time = time
            else:  # "end"
                res[stack.pop()] += time - prev_time + 1
                prev_time = time + 1

        return res

File: LeetCode/636.Exclusive_Time_of_Functions/636.Exclusive_Time_of_Functions.swift
class Solution {
    func exclusiveTime(_ n: Int, _ logs: [String]) -> [Int] {
        var res = Array(repeating: 0, count: n)
        var stack: [Int] = []
        var prev_time = 0

        for log in logs {
            let parts = log.split(separator: ":")
            if let id = Int(parts[0]), let time = Int(parts[2]) {
                let type = parts[1]
                if type == "start" {
                    if let cur = stack.last {
                        res[cur] += time - prev_time
                    }
                    prev_time = time
                    stack.append(id)
                } else {
                    res[stack.last!] += time - prev_time + 1
                    prev_time = time + 1
                    stack.removeLast()
                }
            }
        }

        return res
    }
}
